omegatrade-app/frontend/src/app/components/stock-dashboard/stock-dashboard.component.ts

```diff
@@ -41,141 +46,171 @@ export class StockDashboardComponent implements OnInit {
-    const source = interval(5000);
-    this.subscription = source.subscribe(val => this.updateDashboard());
-  }
-
   updateDashboard() {
     if (this.selectedCompany && this.lastUpdatedTime) {
-      this.restService.getData(`companies/dashboard/${this.selectedCompany}?sId=${this.sId}&date=${this.lastUpdatedTime}`)
+      this.restService.getData(`companies/dashboard/${this.selectedCompany}?date=${this.lastUpdatedTime}`)
         .pipe(take(1))
         .subscribe(
           response => {
             if (response && response.success) {
               const data = response.data.stocks;
               const company = response.data.company;
-              if (company.status == 'COMPLETED' && this.subscription && !this.subscription.closed) {
-                this.subscription && this.subscription.unsubscribe();
+              if (company.status == 'COMPLETED' || company.status === null) {
+                  // Canceling Subscrition when simulation completed
+                  this.clearAllTimeOuts()
               } else {
-                this.stocks = data;
                 if (data && data.length > 0) {
-                  this.lastUpdatedTime = this.stocks[(this.stocks.length - 1)].date;
+                  // updating lastUpdatedtime with current stock data
+                  this.lastUpdatedTime = data[(data.length - 1)].date;
                   this.stock.ref$.subscribe(chart => {
                     for (var i = 0; i < data.length; i++) {
                       chart.series[0].addPoint([data[i].date, parseInt(data[i].currentValue)]);
                     }
                   });
-                }else if(company && company.status == 'PROCESSING'){
-                  console.log('here')
-                  setTimeout(() => {
+                }
+                if(company && company.status == 'PROCESSING'){
+                  /**
+                   * Retrying getStockData in the case of empty stocks at that current time. 
+                   * But it may have data since the staus is in PROCESSING,
+                   * so fetching the datas of running simulation in certain interval.
+                   * 
+                   */
+                   const id = setTimeout(() => {
                     this.updateDashboard();
                   }, 5000);
+                  this.timerIds.push(id);
                 }
               }
             }
           },
           error => {
             if (error && error.error && error.error.message) {
               this.snackBarService.openSnackBar(error.error.message, '');
-              this.ngOnDestroy();
             }
+            this.clearAllTimeOuts();
             this.loader = false;
           });
     }
   }
-
-  changeCompany() {
-    this.getStockData()
+  
+  /**
+   * Function to clear all timeout functions
+   */
+  clearAllTimeOuts() {
+    if(this.timerIds && this.timerIds.length>0){
+      let len = this.timerIds.length;
+      const ids = this.timerIds;
+      while (len > 0) {
```

Comment:
instead of while loop I would suggest to put for loop here

Reply:
this suggestion has been added.